File: claude/harness/steambench_harness/envs/connect4.py
"""Connect Four — a deterministic SteamBench *adversarial* env.

The one category the other arcade games don't cover: a two-player game. You play
a best-of-6 series against a fixed, rule-based opponent (win-if-you-can,
block-if-you-must, else play center-out). The opponent is deterministic and
there's no RNG, so a run is fully determined by your drops alone — trivially
replay-verifiable. Skill (lookahead) translates straight into wins.
"""

from __future__ import annotations

from typing import Optional, Union

from ..protocol import AchievementSpec, ActionSpace, Env, Observation, VerifyMode
from ..registry import register

COLS, ROWS = 7, 6
GAMES = 6  # best-of series
CENTER_ORDER = [3, 2, 4, 1, 5, 0, 6]  # opponent's tie-break preference


def _drop_row(board: list[list[int]], col: int) -> Optional[int]:
    """Lowest empty row in a column, or None if full."""
    for r in range(ROWS - 1, -1, -1):
        if board[r][col] == 0:
            return r
    return None


def _wins(board: list[list[int]], who: int) -> bool:
    for r in range(ROWS):
        for c in range(COLS):
            if board[r][c] != who:
                continue
            for dr, dc in ((0, 1), (1, 0), (1, 1), (1, -1)):
                rr, cc, n = r, c, 0
                while 0 <= rr < ROWS and 0 <= cc < COLS and board[rr][cc] == who:
                    n += 1
                    if n == 4:
                        return True
                    rr += dr
                    cc += dc
    return False


def _legal(board: list[list[int]]) -> list[int]:
    return [c for c in range(COLS) if board[0][c] == 0]


def _winning_move(board: list[list[int]], who: int) -> Optional[int]:
    for c in _legal(board):
        r = _drop_row(board, c)
        board[r][c] = who
        win = _wins(board, who)
        board[r][c] = 0
        if win:
            return c
    return None


@register
class Connect4(Env):
    env_id = "arcade/connect4"
    appid = 9000008
    name = "Connect Four"
    description = (
        "Play a best-of-6 series of Connect Four against a fixed rule-based "
        "opponent. Drop discs to make four in a row; out-think the bot to win."
    )
    verify_mode = VerifyMode.REPLAY
    action_space = ActionSpace([str(c) for c in range(COLS)])
    achievements = [
        AchievementSpec("win_1", "On the Board", "Win a game vs the bot.", 0.7),
        AchievementSpec("win_2", "Got Its Number", "Win 2 games.", 0.45),
        AchievementSpec("win_3", "Majority", "Win 3 games.", 0.28),
        AchievementSpec("win_5", "Dominant", "Win 5 games.", 0.1),
        AchievementSpec("sweep", "Flawless", f"Win all {GAMES} — losing none.", 0.04),
    ]

    def __init__(self) -> None:
        super().__init__()
        self.board: list[list[int]] = [[0] * COLS for _ in range(ROWS)]
        self.game_index = 0
        self.wins = 0
        self.losses = 0
        self.draws = 0

    def reset(self, seed: int = 0) -> Observation:
        self._begin(seed)
        self.game_index = 0
        self.wins = 0
        self.losses = 0
        self.draws = 0
        self._new_game()
        return self._observe(0.0)

    def _new_game(self) -> None:
        self.board = [[0] * COLS for _ in range(ROWS)]
        # Alternate who moves first across the series; opponent-first games open
        # with the opponent's deterministic move.
        if self.game_index % 2 == 1:
            self._opponent_move()

    def _opponent_move(self) -> None:
        win = _winning_move(self.board, 2)
        block = _winning_move(self.board, 1)
        if win is not None:
            col = win
        elif block is not None:
            col = block
        else:
            legal = _legal(self.board)
            col = next((c for c in CENTER_ORDER if c in legal), legal[0] if legal else 0)
        r = _drop_row(self.board, col)
        if r is not None:
            self.board[r][col] = 2

    def _end_game(self, result: str) -> float:
        if result == "win":
            self.wins += 1
        elif result == "loss":
            self.losses += 1
        else:
            self.draws += 1
        self.score = self.wins
        reward = 1.0 if result == "win" else (-1.0 if result == "loss" else 0.0)
        self._check_achievements()
        self.game_index += 1
        if self.game_index >= GAMES:
            self.done = True
        else:
            self._new_game()
        return reward

    def step(self, action: Union[int, str]) -> Observation:
        if self.done:
            return self._observe(0.0)
        col = int(self.action_space.name(action))
        self.steps += 1
        r = _drop_row(self.board, col)
        if r is None:
            return self._observe(0.0)  # illegal (full column) — no-op
        self.board[r][col] = 1
        if _wins(self.board, 1):
            return self._observe(self._end_game("win"))
        if not _legal(self.board):
            return self._observe(self._end_game("draw"))
        # opponent replies
        self._opponent_move()
        if _wins(self.board, 2):
            return self._observe(self._end_game("loss"))
        if not _legal(self.board):
            return self._observe(self._end_game("draw"))
        return self._observe(0.0)

    def _check_achievements(self) -> None:
        for n, aid in [(1, "win_1"), (2, "win_2"), (3, "win_3"), (5, "win_5")]:
            if self.wins >= n:
                self._unlock(aid)
        if self.wins >= GAMES and self.losses == 0:
            self._unlock("sweep")

    def legal_actions(self) -> list[str]:
        return [str(c) for c in _legal(self.board)]

    def render(self) -> str:
        sym = {0: ".", 1: "X", 2: "O"}
        head = f"game {self.game_index + 1}/{GAMES}  W{self.wins}-L{self.losses}-D{self.draws}"
        rows = ["".join(sym[v] for v in row) for row in self.board]
        return head + "\n" + "\n".join(rows)

    def _observe(self, reward: float) -> Observation:
        return Observation(
            step=self.steps,
            state={
                "board": [row[:] for row in self.board],
                "cols": COLS,
                "rows": ROWS,
                "game": self.game_index + 1,
                "games": GAMES,
                "wins": self.wins,
                "losses": self.losses,
                "draws": self.draws,
                "you": 1,
                "opponent": 2,
                "score": self.score,
            },
            text=self.render(),
            legal_actions=self.legal_actions(),
            score=float(self.score),
            done=self.done,
            reward=float(reward),
            info={"wins": self.wins, "losses": self.losses},
        )

File: claude/web/src/lib/arcade/connect4.ts
/**
 * Connect Four — a deterministic SteamBench *adversarial* env.
 *
 * TypeScript mirror of `harness/steambench_harness/envs/connect4.py`. You play a
 * best-of-6 series against a fixed, rule-based opponent (win-if-you-can,
 * block-if-you-must, else play center-out). The opponent is deterministic and
 * there's no RNG, so a run is fully determined by your drops alone — trivially
 * replay-verifiable. Skill (lookahead) translates straight into wins.
 */

import { ActionSpace, Env } from "./base";
import type { AchievementSpec, Action, Observation } from "./types";

const COLS = 7;
const ROWS = 6;
const GAMES = 6; // best-of series
const CENTER_ORDER = [3, 2, 4, 1, 5, 0, 6]; // opponent's tie-break preference

type Board = number[][];

/** Lowest empty row in a column, or null if full. */
function dropRow(board: Board, col: number): number | null {
  for (let r = ROWS - 1; r >= 0; r--) {
    if (board[r][col] === 0) {
      return r;
    }
  }
  return null;
}

function wins(board: Board, who: number): boolean {
  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      if (board[r][c] !== who) {
        continue;
      }
      for (const [dr, dc] of [
        [0, 1],
        [1, 0],
        [1, 1],
        [1, -1],
      ]) {
        let rr = r;
        let cc = c;
        let n = 0;
        while (rr >= 0 && rr < ROWS && cc >= 0 && cc < COLS && board[rr][cc] === who) {
          n += 1;
          if (n === 4) {
            return true;
          }
          rr += dr;
          cc += dc;
        }
      }
    }
  }
  return false;
}

function legal(board: Board): number[] {
  const out: number[] = [];
  for (let c = 0; c < COLS; c++) {
    if (board[0][c] === 0) {
      out.push(c);
    }
  }
  return out;
}

/** First legal col (in 0..COLS order) that makes `who` win, else null. */
function winningMove(board: Board, who: number): number | null {
  for (const c of legal(board)) {
    const r = dropRow(board, c);
    if (r === null) {
      continue;
    }
    board[r][c] = who;
    const win = wins(board, who);
    board[r][c] = 0;
    if (win) {
      return c;
    }
  }
  return null;
}

function emptyBoard(): Board {
  const board: Board = [];
  for (let r = 0; r < ROWS; r++) {
    board.push(new Array<number>(COLS).fill(0));
  }
  return board;
}

export class Connect4 extends Env {
  static env_id = "arcade/connect4";
  static appid = 9000008;
  static displayName = "Connect Four";
  static description =
    "Play a best-of-6 series of Connect Four against a fixed rule-based " +
    "opponent. Drop discs to make four in a row; out-think the bot to win.";
  static verify_mode = "replay" as const;
  static action_space = new ActionSpace(
    Array.from({ length: COLS }, (_, c) => String(c)),
  );
  static achievements: AchievementSpec[] = [
    { id: "win_1", name: "On the Board", description: "Win a game vs the bot.", rarity_hint: 0.7 },
    { id: "win_2", name: "Got Its Number", description: "Win 2 games.", rarity_hint: 0.45 },
    { id: "win_3", name: "Majority", description: "Win 3 games.", rarity_hint: 0.28 },
    { id: "win_5", name: "Dominant", description: "Win 5 games.", rarity_hint: 0.1 },
    { id: "sweep", name: "Flawless", description: `Win all ${GAMES} — losing none.`, rarity_hint: 0.04 },
  ];

  board: Board;
  gameIndex: number;
  wins: number;
  losses: number;
  draws: number;

  constructor() {
    super();
    this.board = emptyBoard();
    this.gameIndex = 0;
    this.wins = 0;
    this.losses = 0;
    this.draws = 0;
  }

  // ---- lifecycle --------------------------------------------------------- //

  reset(seed = 0): Observation {
    this.begin(seed); // seed unused (no RNG) but keeps the contract
    this.gameIndex = 0;
    this.wins = 0;
    this.losses = 0;
    this.draws = 0;
    this.newGame();
    return this.observe(0.0);
  }

  private newGame(): void {
    this.board = emptyBoard();
    // Alternate who moves first across the series; opponent-first games open
    // with the opponent's deterministic move.
    if (this.gameIndex % 2 === 1) {
      this.opponentMove();
    }
  }

  private opponentMove(): void {
    const win = winningMove(this.board, 2);
    const block = winningMove(this.board, 1);
    let col: number;
    if (win !== null) {
      col = win;
    } else if (block !== null) {
      col = block;
    } else {
      const lg = legal(this.board);
      const centered = CENTER_ORDER.find((c) => lg.includes(c));
      col = centered !== undefined ? centered : lg.length > 0 ? lg[0] : 0;
    }
    const r = dropRow(this.board, col);
    if (r !== null) {
      this.board[r][col] = 2;
    }
  }

  private endGame(result: string): number {
    if (result === "win") {
      this.wins += 1;
    } else if (result === "loss") {
      this.losses += 1;
    } else {
      this.draws += 1;
    }
    this.score = this.wins;
    const reward = result === "win" ? 1.0 : result === "loss" ? -1.0 : 0.0;
    this.checkAchievements();
    this.gameIndex += 1;
    if (this.gameIndex >= GAMES) {
      this.done = true;
    } else {
      this.newGame();
    }
    return reward;
  }

  step(action: Action): Observation {
    if (this.done) {
      return this.observe(0.0);
    }
    const col = parseInt(this.actionSpace.name(action), 10);
    this.steps += 1;
    const r = dropRow(this.board, col);
    if (r === null) {
      return this.observe(0.0); // illegal (full column) — no-op
    }
    this.board[r][col] = 1;
    if (wins(this.board, 1)) {
      return this.observe(this.endGame("win"));
    }
    if (legal(this.board).length === 0) {
      return this.observe(this.endGame("draw"));
    }
    // opponent replies
    this.opponentMove();
    if (wins(this.board, 2)) {
      return this.observe(this.endGame("loss"));
    }
    if (legal(this.board).length === 0) {
      return this.observe(this.endGame("draw"));
    }
    return this.observe(0.0);
  }

  private checkAchievements(): void {
    for (const [n, aid] of [
      [1, "win_1"],
      [2, "win_2"],
      [3, "win_3"],
      [5, "win_5"],
    ] as Array<[number, string]>) {
      if (this.wins >= n) {
        this.unlock(aid);
      }
    }
    if (this.wins >= GAMES && this.losses === 0) {
      this.unlock("sweep");
    }
  }

  // ---- introspection ----------------------------------------------------- //

  legalActions(): string[] {
    return legal(this.board).map((c) => String(c));
  }

  render(): string {
    const sym: Record<number, string> = { 0: ".", 1: "X", 2: "O" };
    const head = `game ${this.gameIndex + 1}/${GAMES}  W${this.wins}-L${this.losses}-D${this.draws}`;
    const rows = this.board.map((row) => row.map((v) => sym[v]).join(""));
    return head + "\n" + rows.join("\n");
  }

  private observe(reward: number): Observation {
    return {
      step: this.steps,
      state: {
        board: this.board.map((row) => row.slice()),
        cols: COLS,
        rows: ROWS,
        game: this.gameIndex + 1,
        games: GAMES,
        wins: this.wins,
        losses: this.losses,
        draws: this.draws,
        you: 1,
        opponent: 2,
        score: this.score,
      },
      text: this.render(),
      frame: null,
      legal_actions: this.legalActions(),
      score: this.score,
      done: this.done,
      reward: reward,
      info: { wins: this.wins, losses: this.losses },
    };
  }
}
